membership-attribute-service/app/services/AttributesFromZuora.scala

```diff
@@ -43,64 +42,65 @@ object AttributesFromZuora extends LoggingWithLogstashFields {
     }
 
 
-    val attributesFromDynamo = dynamoAttributeGetter(identityId)
+    val attributesFromDynamo = dynamoAttributeService.get(identityId)
 
-    val attributesFromZuora = for {
-      dynamoAttributes <- attributesFromDynamo
-      zuoraAttributes <- attributesDisjunction.run.map {
+    val attributesFromZuora: Future[Disjunction[String, Option[Attributes]]] = attributesDisjunction.run.map {
         _.leftMap { errorMsg =>
           log.error(s"Tried to get Attributes for $identityId but failed with $errorMsg")
-          dynamoAttributes
-        }.fold(_ => dynamoAttributes, identity)
+          errorMsg
       }
-    } yield zuoraAttributes
-
-    val shouldSkipUpdate = dynamoAndZuoraAgree(attributesFromDynamo, attributesFromZuora, identityId)
-
-    shouldSkipUpdate.onFailure {
-      case error => log.warn(s"Tried to compare attributes for $identityId but then ${error.getMessage}", error)
     }
 
-    val zuoraAttributesWithAdfree = attributesWithFlagFromDynamo(attributesFromZuora, attributesFromDynamo)
+    val attributesFromZuoraOrDynamoFallback: Future[(String, Option[Attributes])] = for {
+      dynamoAttributes <- attributesFromDynamo
+      attributesFromZuora <- attributesFromZuora
+    } yield attributesFromZuora.fold(_ => ("Dynamo", dynamoAttributes), ("Zuora", _))
+
+    attributesFromZuoraOrDynamoFallback flatMap { attributesFromSomewhere =>
+      val (fromWhere: String, attributes: Option[Attributes]) = attributesFromSomewhere
 
-    attributesAfterDynamoUpdate(identityId, shouldSkipUpdate, zuoraAttributesWithAdfree, dynamoAttributeUpdater)
+      if(fromWhere == "Zuora") {
+        attributesFromDynamo flatMap { (dynamoAttributes: Option[Attributes]) =>
+          val shouldSkipUpdate: Boolean = dynamoAndZuoraAgree(dynamoAttributes, attributes, identityId)
+          val zuoraAttributesWithAdfree: Option[Attributes] = attributesWithFlagFromDynamo(attributes, dynamoAttributes)
+          attributesAfterDynamoUpdate(identityId, shouldSkipUpdate, zuoraAttributesWithAdfree, dynamoAttributeService)
```

Comment:
We could consider running this cache update in a separate thread, and then just returning zuoraAttributesWithAdFree regardless of the result of the cache update?

Reply:
If we did that, we'd probably want to rename attributesAfterDynamoUpdate to something like 'updateDynamo' or 'updateCache'...